### PR TITLE
Bug 1981957: various fixes on top of v1.0.47 to satisfy regression testing

### PR DESCRIPTION
**Pull request by gabemontero**
This will need https://github.com/openshift/origin/pull/26336 to merge in order for the imagestreamtag pod template test to pass in e2e-aws-jenkins

See https://github.com/openshift/jenkins/pull/1297 for the tl;dr saga of how/why I identified these changes

Unfortunately there are some whitespace editor formating diffs between my editor (intellij) and whatever was last used.  If I am going to end up contributing more often to this repo, I can work with @akram to sort that out.

But in the meantime be sure the view with `?w=1` when looking at the file diffs.

i.e. https://github.com/openshift/jenkins-sync-plugin/pull/389/files?w=1

@openshift/openshift-team-build-api @openshift/team-devtools-jenkins

### CODE REVIEW

File: src/main/resources/io/fabric8/jenkins/openshiftsync/GlobalPluginConfiguration/config.jelly
Comment (akram) [inline src/main/resources/io/fabric8/jenkins/openshiftsync/GlobalPluginConfiguration/config.jelly:95]: if we remove this, we need to remove all the `validate` method.
But, the validation method is quite informative for sizing purposing and configuration of the cluster mode

File: src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java:427]: for null protection, can we use 
```
LEGACY_IMAGESTREAM_AGENT_LABEL_VALUE.equals(map.get(IMAGESTREAM_AGENT_LABEL)) 
```
and extract `map.get(IMAGESTREAM_AGENT_LABEL)`  into a variable.
?

**Review by akram — CHANGES_REQUESTED** [review]

File: src/main/resources/io/fabric8/jenkins/openshiftsync/GlobalPluginConfiguration/config.jelly
Comment (gabemontero) [reply src/main/resources/io/fabric8/jenkins/openshiftsync/GlobalPluginConfiguration/config.jelly:95]: I'm fine leaving the validation in if we can figure out how to use it *AND* how to have the fields in question actually show up on the config panel.

They currently do not appear.

Any idea why that is the case @akram ?

File: src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java
Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java:427]: Sure thing.  Will work on that shortly.

Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java]: update pushed

**Comment by akram** [comment]
/lgtm

needs https://github.com/openshift/origin/pull/26336 before getting able to merge

**Comment by gabemontero** [comment]
/hold

aside from waiting on openshift/origin PR, my attempt at accepting multiple values for the pod template label is not working for imagestreams

so I'll either need to update that commit, or revert it if I can't address properly in time for customers waiting on @akram 's scaling fix

**Comment by gabemontero** [comment]
/hold cancel

per conversation with @adambkaplan and @akram on slack, even setting up separate informers for each key/value pair for the imagestream pod template label did not do the trick.

Perhaps an issue with the fabric shared informer??

I've reverted the commit for that in this PR.  We'll have to pursue that in a more dedicated fashion, in a separate tracking item / PR.

I also notice the config maps were not employing any label filter.  I'm going to try and tackle that with this PR.

**Comment by akram** [comment]
The approach with 2 informers will make the scalability issue worse, so focusing on one informer per resource type/namespace is the solution.

/lgtm

**Comment by gabemontero** [comment]
Looks like CI for the plugin and building the jenkins image is not quite right now.  See this in one of the failure's Jenkins log:

```
2021-07-20 09:01:27 SEVERE  jenkins.InitReactorRunner$1 onTaskFailed Failed Loading plugin OpenShift Sync v1.0.47-SNAPSHOT (private-964320a0-root) (openshift-sync)
java.io.IOException: Failed to load: OpenShift Sync (1.0.47-SNAPSHOT (private-964320a0-root))
 - Update required: Jackson 2 API Plugin (2.12.1) to be updated to 2.12.3 or higher
 - Update required: Kubernetes Client API Plugin (4.13.3-1) to be updated to 5.4.1 or higher
 - Update required: Kubernetes plugin (1.29.7) to be updated to 1.30.0 or higher
	at hudson.PluginWrapper.resolvePluginDependencies(PluginWrapper.java:963)
	at hudson.PluginManager$2$1$1.run(PluginManager.java:550)
	at org.jvnet.hudson.reactor.TaskGraphBuilder$TaskImpl.run(TaskGraphBuilder.java:169)
	at org.jvnet.hudson.reactor.Reactor.runTask(Reactor.java:296)
	at jenkins.model.Jenkins$5.runTask(Jenkins.java:1129)
	at org.jvnet.hudson.reactor.Reactor$2.run(Reactor.java:214)
	at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
	at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:68)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

I didn't think we had to use the "snapshot" / "PR-Testing" form any more.  And I don't remember that occurring with my recent client plugin test case work.

Looks like either an update to this PR to sort out snapshot short term, or analysis of openshift/release (and maybe the Dockerfile here) to sort out building the jenkins image from CI, is in order.

**Comment by gabemontero** [comment]
Or maybe we need our base jenkins image updated to match the new mvn dependencies in our sync plugin pom.xml from what 1.0.47 introduced?

That sounds plausible ... also, I realized that until we cut an official version of sync plugin, or PR builds will say "snapshot".  Also, I found the log that says we were not going down the PR path.

That would mean splitting out the bumps from https://github.com/openshift/jenkins/pull/1297 ... the k8s / jackson updates in one PR.  Hopefully v1.0.46 works with those.   Once that merges, our tests here pass.

Or we override tests here (again).

Sigh.

I'll get started on the new openshift/jenkins PR, and cross my fingers.

@openshift/openshift-team-build-api  @openshift/team-devtools-jenkins FYI

**Comment by gabemontero** [comment]
> Or maybe we need our base jenkins image updated to match the new mvn dependencies in our sync plugin pom.xml from what 1.0.47 introduced?
> 
> That sounds plausible ... also, I realized that until we cut an official version of sync plugin, or PR builds will say "snapshot". Also, I found the log that says we were not going down the PR path.
> 
> That would mean splitting out the bumps from [openshift/jenkins#1297](https://github.com/openshift/jenkins/pull/1297) ... the k8s / jackson updates in one PR. Hopefully v1.0.46 works with those. Once that merges, our tests here pass.

No dice ^^.  See https://github.com/openshift/jenkins/pull/1308#issuecomment-883498461

I'll see if I can factor in a dependent plugin update in this repo's e2e job.  Otherwise we are overriding @openshift/openshift-team-build-api @openshift/team-devtools-jenkins and depending on my local running of the e2e's


> 
> Or we override tests here (again).
> 
> Sigh.
> 
> I'll get started on the new openshift/jenkins PR, and cross my fingers.
> 
> @openshift/openshift-team-build-api @openshift/team-devtools-jenkins FYI

**Comment by gabemontero** [comment]
@openshift/openshift-team-build-api  @openshift/team-devtools-jenkins Victory!!!

green e2e

just need a lgtm

**Comment by gabemontero** [comment]
btw don't feel we should squash the commits here ... solved multiple problems with this one to get us green

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildManager.java
Comment (adambkaplan) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildManager.java:299]: What does this function even do? It initializes a set of builds to remove, which a string convention that doesn't seem amenable to anything (namespace + name are concatenated with no spearator). We then call `iter.next()` twice in the same loop without calling `iter.hasNext()` beforehand. We then do nothing with that set.

**Review by adambkaplan — CHANGES_REQUESTED** [review]
It isn't clear to me that `removeBuildsForDeletedBCFromNoBCList actually does anything.

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildManager.java
Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/BuildManager.java:299]: I found the link where it says sequential events can be assumed:  https://github.com/kubernetes/client-go/blob/master/tools/cache/shared_informer.go#L133-L136

So my edge case should not happen.  The build will get flushed by the add bc event before any delete bc event comes.

Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/BuildManager.java]: method removed @adambkaplan

**Review by adambkaplan — APPROVED** [review]
/lgtm

**Comment by gabemontero** [comment]
@akram I have pushed the 5 commits from this PR to https://github.com/jenkinsci/openshift-sync-plugin/commits/master

**Comment by akram** [comment]
ART request: https://issues.redhat.com/browse/ART-3173
cc @jitendar-singh

**Pull request merged**
